lib/Sema/MiscDiagnostics.cpp

```diff
@@ -105,13 +105,35 @@ static void diagSyntacticUseRestrictions(const Expr *E, const DeclContext *DC,
     SmallPtrSet<DeclRefExpr*, 4> AlreadyDiagnosedBitCasts;
 
     bool IsExprStmt;
+    bool ShouldDiagnoseTypeDefaultedCollectionAsWarning;
```

Comment:
I don't think this is necessary because `ASTWalker` has a `Parent` which could be used to determine this when `CollectionExpr` is found - if there is no parent or it's semantically meaningless - diagnose as an error, otherwise diagnose as a warning.

Reply:
What if there's sth semantically meaningful surrounding a semantically meaningless parent, like
 ```swift
let _ = (true, ([1, "Swift"]))
```
? That would have to be a warning.

Reply:
Too bad ASTWalker doesn't carry a depth but it should be easy to add that and increase for each semantically significant expression...

Reply:
Right, I was thinking about a `hasReachedValueProvidingExpr` flag.

Reply:
Something like that sounds good to me!

Reply:
@xedin I just pushed new changes. Could you take a look?

Reply:
Absolutely!